src/Vendr.PaymentProviders.Worldpay/WorldpayPaymentProvider.cs

```diff
@@ -1,38 +1,139 @@
 using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
 using System.Web;
 using System.Web.Mvc;
 using Vendr.Core;
+using Vendr.Core.Logging;
 using Vendr.Core.Models;
+using Vendr.Core.Services;
 using Vendr.Core.Web.Api;
 using Vendr.Core.Web.PaymentProviders;
+using Vendr.PaymentProviders.Worldpay.Helpers;
 
-namespace Vendr.PaymentProviders
+namespace Vendr.PaymentProviders.Worldpay
 {
-    [PaymentProvider("worldpay", "Worldpay", "Worldpay payment provider", Icon = "icon-invoice")]
+    [PaymentProvider("worldpay", "Worldpay", "Worldpay payment provider", Icon = "icon-credit-card")]
     public class WorldpayPaymentProvider : PaymentProviderBase<WorldpaySettings>
     {
-        public WorldpayPaymentProvider(VendrContext vendr)
-            : base(vendr)
-        { }
+        private readonly ILogger _logger;
 
-        public override bool FinalizeAtContinueUrl => true;
+        public WorldpayPaymentProvider(VendrContext vendr, ILogger logger) : base(vendr)
+        {
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+        }
+
+        public override bool FinalizeAtContinueUrl => false;
 
         public override PaymentFormResult GenerateForm(OrderReadOnly order, string continueUrl, string cancelUrl, string callbackUrl, WorldpaySettings settings)
         {
-            return new PaymentFormResult()
+            try
+            {
+                if (settings.VerboseLogging)
+                {
+                    _logger.Info<WorldpayPaymentProvider>($"GenerateForm method called for cart {order.CartNumber}");
+                }
+
+                var url = settings.Mode.ToLower() == "live" ? settings.LiveUrl : settings.TestUrl;
+                var form = new PaymentForm(url, FormMethod.Post);
+
+                settings.InstallId.MustNotBeNull("settings.InstallId");
+                settings.TestModeNumber.MustNotBeNull("settings.TestModeNumber");
+                settings.AuthMode.MustNotBeNull("settings.AuthMode");
+                settings.OrderPropertyBillingFirstName.MustNotBeNull("settings.OrderPropertyBillingFirstName");
+                settings.OrderPropertyBillingLastName.MustNotBeNull("settings.OrderPropertyBillingLastName");
+
+                var firstname = order.Properties[settings.OrderPropertyBillingFirstName];
+                var surname = order.Properties[settings.OrderPropertyBillingLastName];
```

Comment:
Vendr should already know the customer first name + last name + email so I wonder if these should be accessed from `order.CustomerInfo.FirstName` and `order.CustomerInfo.LastName`? Or may allow those to be null in the settings and fallback to these fields. In this case, you might need to drop the null check above.

Reply:
We've updated it to use CustomerInfo unless there is a setting added, removed null check